### PR TITLE
In Horizontal Pod Autoscaler Controller when scaling on multiple metrics, handle invalid metrics.

### LINKED ISSUE #61007: HPA refuses to scale if any custom metric is missing
> /kind bug

**What happened**:

I setup custom metrics based autoscaling with multiple metrics.  One of those metrics was not available.  HPA reports all metrics as "unknown" (even the ones that are still available) and refuses to operate.

```
Metrics:                                               ( current / target )
  "sockjs_sessions_current" on pods:                   <unknown> / 500
  "ddp_method_calls" on pods:                          <unknown> / 25
  "http_requests" on pods:                             <unknown> / 25
  resource cpu on pods  (as a percentage of request):  <unknown> / 60%
Conditions:
  Type           Status  Reason               Message
  ----           ------  ------               -------
  AbleToScale    True    SucceededGetScale    the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetPodsMetric  the HPA was unable to compute the replica count: unable to get metric ddp_method_calls: no metrics returned from custom metrics API
Events:
  Type     Reason               Age                  From                       Message
  ----     ------               ----                 ----                       -------
  Warning  FailedGetPodsMetric  1m (x1052 over 17h)  horizontal-pod-autoscaler  unable to get metric ddp_method_calls: no metrics returned from custom metrics API
```

**What you expected to happen**:

Pods would be automatically scaled based on metrics that were available, ignoring missing metrics (treating them as zero).

A partial failure in the metrics system should not prevent auto-scaling from proceeding with the data that it has.  For example, CPU % metrics come from the k8s metrics server whereas the other metrics come from the prometheus adapter.  If the prometheus adapter goes away we can still use the CPU metric as a lower bound on the number of replicas.

If you consider, e.g. https://github.com/kubernetes/kubernetes/blob/e99ec245958f82acb2404f8597844d62b8f459c9/pkg/controller/podautoscaler/horizontal.go#L242 , instead of aborting the whole metric gathering process there, it could have stored a placeholder or zero metric value when it fails to fetch the metric.

**How to reproduce it (as minimally and precisely as possible)**:

Setup a custom metrics server and tell HPA to scale based on several metrics, one of which is not actually available.

**Environment**:

kops 1.8 on AWS
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.8", GitCommit:"2f73858c9e6ede659d6828fe5a1862a48034a0fd", GitTreeState:"clean", BuildDate:"2018-02-09T21:23:25Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

### PR DESCRIPTION
**Pull request by bskiba**
**What this PR does / why we need it**:
In Horizontal Pod Autoscaler Controller handle case when scaling on multiple metrics and some of them are invalid (misconfigured or missing).
If all metrics are missing, return error and set HPA condition to error from first invalid metric.
If some metrics are missing and some valid, on a scale up ignore missing metrics, on a scale down return an error and set HPA ScalingActive condition to reason coming from first invalid metric.

Reasoning behind this solution: if one metric is unavailable but the other says we should scale up, it is safe to scale up as the missing metric could only tell us to scale even more. If one metric is unavailable (can be transient unavailability) it is not safe to scale up as the missing metric can be the one telling us to keep the replicas count at the current level (or even scale up).

When multiple metrics are unavailable, I set the HPA ScalingActive condition to reason coming from first invalid metric with a reasouning that if you have multiple problems, you will start eliminating them one by one. I can also consider adding a new condition to describe the multiple metrics unavailable situation.

Metric statuses are correctly set on metrics that are available. Events for unavailable metrics are sent as before.

Tested this by 
1. creating hpa with two metric sources, one metric unavailable, second triggering scale up - successfully triggered scaled up for the valid metric. 
2. creating hpa with two metric sources, one metric unavailable, second triggering scale down - verified no scale down.

**Which issue(s) this PR fixes**:
Fixes #61007

**Release note**:
```
In Horizontal Pod Autoscaler Controller handle case when scaling on multiple metrics and some of them are invalid (misconfigured or missing).
```

### CODE REVIEW

**Comment by bskiba** [comment]
@DirectXMan12 @MaciekPytel Would you have time to have a look at this?

**Comment by yastij** [comment]
/remove-lifecycle rotten
@kubernetes/sig-autoscaling-misc

File: pkg/controller/podautoscaler/horizontal.go [hunk 1]
```diff
@@ -200,160 +200,238 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
...
 		}
+	}
 
-		var replicaCountProposal int32
-		var utilizationProposal int64
-		var timestampProposal time.Time
-		var metricNameProposal string
+	// If all metrics are invalid or some are invalid and we would scale down,
+	// return error and set condition on hpa based on first invalid metric.
+	if invalidMetricsCount >= len(metricSpecs) || (invalidMetricsCount > 0 && replicas < currentReplicas) {
```
Comment (gjtempleton) [inline pkg/controller/podautoscaler/horizontal.go]: It feels like the naming of `replicas` here is a bit misleading now (although the name still makes sense at return time) as it's only the current proposed new replica count at the time of this comparison.

It took me a few scans to figure out what it actually corresponded to at the point where this comparison is done.

Comment (matthewejohnson711) [reply pkg/controller/podautoscaler/horizontal.go]: Must work so much better.

**Comment by gjtempleton** [comment]
Think this can now be close as it's been superseded by #78503

**Pull request closed without merging**